### DISCUSSION TITLE
Component proposal: List

### DISCUSSION THREAD

**Discussion by janhassel** [carbon-design-system/carbon#10263] [Ideas]
# List

## Overview

A list is a structured way to display grouped information. The individual list items can contain static / read-only information as well as being interactive. A list is preceded by the list header, which always contains the title for a group of list items and can optionally hold additional, supporting information or interactive elements.

### When to use

- Items can be logically grouped
- Items follow the same structure
- Items are simple and can be displayed in a compact format

### When not to use
- Items don't have a relation to each other
- Items vary in format and structure
- Items contain complex interactions or should be displayed in larger sizes

## Variants

| Variant | Purpose |
| :- | :- |
| Default | When the list is placed as a normal part of the user interface, like on a tile or in a sidebar. |
| Compact | When the list is placed within a temporary context, like a popover. The list must span the entire width of the parenting element and the parenting element must either have an outer border or shadow to visually separate its content from the rest of the user interface. |

### Default

![Default variant](https://user-images.githubusercontent.com/28265588/145433090-b10186be-37c9-4097-b135-c3dfb06cfe37.png)

### Compact

![Compact variant](https://user-images.githubusercontent.com/28265588/145433131-e52aa86a-0107-4807-bdc5-ee29b77b7723.png)

### Anatomy

![Anatomy of the List component](https://user-images.githubusercontent.com/28265588/145433183-bee4fd41-9c1e-409b-8cff-20cc804e0512.png)

1. **List header:** Is used to group the succeeding list items.
2. **List title:** Gives the list group a descriptive name.
3. **Additional list header content (optional):** Can contain supporting information or interactive elements for this list group.
4. **List item:** An entry of the list. The content is completely flexible.

## Universal behaviours

*The behaviors listed in this section are universal across all of the types. For behaviors that are unique to each type, see each of the component type sections below.*

### States

- If a list item is interactive, it should follow the interactive states of the [Tile component](https://www.carbondesignsystem.com/components/tile/usage/) in regards to hover, active and focus.

### Interactions

![Example of how the list component behaves when scrolling](https://user-images.githubusercontent.com/28265588/145433313-48710f3d-adda-40ef-98e3-62d0e4589f98.gif)

- A list header is always visible for as long as its associated list items are visible on the screen. To achieve this, the list header sticks to the top of its parenting element and only allows itself to be pushed by a succeeding list header or a new element which is scrolled into view.
- In case a list item is interactive or has interactive elements, each interaction should be a tabstop. Keep the number of interactions to a minimum. *Remember: if a list item is too cmplex or benefits from a larger area, the list component is probably not the right choice. Consider using headings to organize your content.*

## Style

- Ensure that all list items within a list are the same height to avoid visual discordance

---

# Examples

## Default variant
![Example of the default variant](https://user-images.githubusercontent.com/28265588/145433458-5cb70b7b-5eca-4f72-a455-947911a0c68d.png)

## Default variant
![Example of the compact variant](https://user-images.githubusercontent.com/28265588/145433473-f92acef6-58a3-45de-afc9-7babbe8b97b9.png)

*To anyone on the Carbon team, please feel free to reach out via Slack so we can talk about more use cases and examples I can't post here. Also, I'd be more than happy to share the interactiv prototypes.*
@aagonzales @joshblack @laurenmrice @sstrubberg 

# Additional notes

- This component might need a better name to differenatiate it from `UnorderedList` and `OrderedList` (and their shared `ListItem` child component).
- The "Compact" variant name might be misleading and could lead to issues when the component is expanded to support multiple sizes (like `sm`, `md`, `lg`).

**Comment by aagonzales** [comment]
I've been thinking about what the differences are between your list (which is great) and the existing [simple structured list](https://react.carbondesignsystem.com/?path=/story/components-structuredlist--simple&knob-Active%20(active)=true&knob-Description%20(description)=Active%20loading%20indicator&knob-Form%20validation%20UI%20content%20(invalidText)=Number%20is%20not%20valid&knob-Helper%20text%20(helperText)=Optional%20helper%20text.&knob-Label%20(label)=NumberInput%20label&knob-Maximum%20value%20(max)=100&knob-Minimum%20value%20(min)=0&knob-Number%20input%20arrow%20icon%20translation%20IDs%20(for%20translateWithId%20callback)={%22increment.number%22:%22Increment%20number%22,%22decrement.number%22:%22Decrement%20number%22}&knob-Show%20form%20validation%20UI%20(invalid)=true&knob-Small%20(small)=true&knob-Step%20of%20up/down%20arrow%20(step)=10&knob-Value%20(value)=50&knob-Warning%20state%20text%20(warnText)=A%20high%20threshold%20may%20impact%20performance).

Some of the differences I'm seeing are:
- Jan's list only has 1 column header, not multiple columns. (Though can have sub-items on the right side in addition to the left side).
- Jan's list can have inline actions instead of row actions like structured list. 
- The example shown above are all flush to a containers edge, structured list can sit flat on a page without a structure. 
- Jan's list is more a traditional list vs structured list is more table like.

I'm basically just trying to figure out if this is a wholesale new component or if its a new variant of either [list](https://carbondesignsystem.com/components/list/usage/) or [structured list](https://carbondesignsystem.com/components/structured-list/usage/) components.

**Reply by janhassel** [reply]
I think a big difference and probably a reason to treat it as a separate component is that the individual list items are fully customizable by the team. It's more of a grouping design pattern than an enumeration of things I believe.

From the current research, the only aspects the list items should always share regardless of the actual implementtion by the team are:
- Background
- Dividing line underneath at the bottom
- If interactive: hover, focus and active state

**Comment by joshblack** [comment]
Looks great to me 🎉

**Comment by janhassel** [comment]
I think to move this forward, we'd need to...
1. define a name for the component (and its two variants)
2. create design specs for spacing, color and typography (I'd be glad to share the specs I've used so far for above mockups, probably don't need to start from scratch)
3. build out an experimental version of this component (I can definitely help out here as I have a coded prototype of this and could work on porting it over)

@aagonzales What do you think?

**Comment by dianasanborn** [comment]
We have a [list component](https://pages.github.ibm.com/ai-applications/design/components/list/overview) within AI Apps that's used for the exact same use cases above and more. 

Note: The guidance we have on the site was carried over from the former IoT PAL, and needs to be fully revisited to follow the Carbon guidance structure and also evolved design and use cases, but you can get an idea of the additional variations we have. 

You can see how it gets used in other patterns, like a [list builder](https://pages.github.ibm.com/ai-applications/design/patterns/listbuilder/usage). This is a [link](https://next.carbon-addons-iot-react.com/?path=/story/1-watson-iot-list-list--basic-single-column-with-search) to our coded implementation of it as well.

Would love to contribute and collaborate on this as well if it's being picked up at the Carbon level! We have a lot of use cases for it.

**Comment by laurenmrice** [comment]
# Contained list
- **Code:** `Contained list` will be it's own component the the React story book.
- **Design:** `Contained list` will be housed under an overarching `List` category with Unordered list, Ordered list, and Structured list in docs.
- Names of Variants 1 and 2 are still TBD.

------

## Variant 1 (color and type tokens)
![Contained list—Variant 1 (color and type tokens)](https://user-images.githubusercontent.com/43969356/179804674-5b198f03-0f77-4ad1-be1d-be5157fc882e.png)


## Variant 1 (specs)
![Contained list—Variant 1 (specs)](https://user-images.githubusercontent.com/43969356/179804443-c25713d2-cd7c-4664-8215-33fa042de4ae.png)

------

## Variant 2 (color and type tokens)
![type tokens)](https://user-images.githubusercontent.com/43969356/179805271-68b7415f-e481-4970-a1f3-887429a3f8d6.png)

## Variant 2 (specs)
![Contained list—Variant 2 (specs)](https://user-images.githubusercontent.com/43969356/179805383-bd647381-a802-44ae-abf3-9fa6f65cfef4.png)
